### DISCUSSION TITLE
Why are elements inside a page not updating?

### DISCUSSION THREAD

**Discussion by angelhdzmultimedia** [zauberzeug/nicegui#466] [Q&A] [answered]
I have a page with a column and inside some labels,  a To Do app to add tasks.

This code works:

```py
from nicegui import ui
from datetime import datetime

class Task:
  def __init__(self, label: str):
    self.label: str = label
    self.id: str = datetime.now()

tasks = [Task(
  label = 'Task 1'
)]

class Ref:
  def __init__(self, value = None):
    self.value = value

def ref(value):
  return Ref(value)

newTask = ref('')

def addTask():
  tasks.append(Task(
    label = newTask.value
  ))
  updateTasks()
  newTask.value = ''



tasksList = None

def updateTasks():
  global tasksList, main
  tasksList and tasksList.clear()
  
  with tasksList:
    for task in tasks:
      print(task)
      ui.label(task.label)
    
  #main.update()
  
 

main = None


""" @ui.page('/', dark=True)
def index():
  global tasksList, main """
with ui.header(elevated=True).classes('bg-blue'):
  ui.button().props('icon="menu" color="blue"')

main = ui.column()
with main:
  ui.label('Tasks')
  tasksList = ui.column()
  updateTasks()

  ui.input('Task').bind_value(newTask, 'value')
  ui.button().props('icon="add"').classes('bg-blue').on('click', addTask)

ui.run()
```

The labels are added in the column and the column is updated automatically.

But the moment I put everything inside a @ui.page, I have to manually call .update() on the column that holds the label otherwise the column is empty:

```py
from nicegui import ui
from datetime import datetime

class Task:
  def __init__(self, label: str):
    self.label: str = label
    self.id: str = datetime.now()

tasks = [Task(
  label = 'Task 1'
)]

class Ref:
  def __init__(self, value = None):
    self.value = value

def ref(value):
  return Ref(value)

newTask = ref('')

def addTask():
  tasks.append(Task(
    label = newTask.value
  ))
  updateTasks()
  newTask.value = ''



tasksList = None

def updateTasks():
  global tasksList, main
  tasksList and tasksList.clear()
  
  with tasksList:
    for task in tasks:
      print(task)
      ui.label(task.label)
    
  main.update()
  
 

main = None


@ui.page('/', dark=True)
def index():
  global tasksList, main
  with ui.header(elevated=True).classes('bg-blue'):
    ui.button().props('icon="menu" color="blue"')
  
  main = ui.column()
  with main:
    ui.label('Tasks')
    tasksList = ui.column()
    updateTasks()

    ui.input('Task').bind_value(newTask, 'value')
    ui.button().props('icon="add"').classes('bg-blue').on('click', addTask)

ui.run()
```

**Comment by angelhdzmultimedia** [comment]
Yeah, sorry about that. I've been submitting issues in repos without a discusions tab, so I still don't get used to it.

**Reply by falkoschindler** [reply]
No worries. We're here to keep it organized. 🧹 😉

**Comment by falkoschindler** [comment, marked as answer]
I don't understand. The UI in the second code example seems to work as expected. And you are not _calling_ `main.update` (there are not brackets). So you can simply remove this line.

**Reply by angelhdzmultimedia** [reply]
Sorry about that too, edited the code. In the process of copying the code, I really have no idea how it lost the brackets haha.

So the thing is:

- Column inside @ui.page must be updated with .update() method.
- Colum outside of @ui.page updates automatically without calling .update().

Probably I'm doing something wrong. Been in an intensive remote assistance rally last couple of days helping someone with a game engine and my head spins a lot.

YouTube video: https://youtu.be/lJt7QifUobY

**Reply by angelhdzmultimedia** [reply]
I think I made some discovery:

I removed the `main.update()` call when the elements are inside the @ui.page, and noticed that after refreshing the browser, 
then the column started to get updated.

So it's not an issue with the UI itself, but maybe with @ui.page and how pages are rendered from the server side? Or maybe a cache issue?


https://user-images.githubusercontent.com/10191085/223486618-5b99eca1-90a6-48cc-9a06-995d696a74cf.mp4

**Reply by falkoschindler** [reply]
When I copy your code, it's working flawlessly. Maybe the socket connection does not come up correctly after auto-reload? I'd suggest to boil it down, remove unnecessary parts, and look when it works or breaks.

**Reply by angelhdzmultimedia** [reply]
What a strange day! Haha! This issue is related to my other issue with fonts on the official website of nicegui!!! https://github.com/zauberzeug/nicegui/issues/462#issuecomment-1458373449

There's a problem with Chrome, but strangely only with nicegui related stuff!

I tested the code in Firefox, and like in your case, it worked flawlessly!

**Edit:** I'm going to uninstall Chrome, restart Windows 10, and reinstall fresh.

**Reply by angelhdzmultimedia** [reply]
### Ugh! False alarm. 
Started happening on Firefox too.


https://user-images.githubusercontent.com/10191085/223492205-050726d8-05df-44f9-bf64-a59553333a1a.mp4

**Reply by angelhdzmultimedia** [reply]
Ok ok ok. Thank to your tip about "sockets", I h ad an idea.

I closed all the previously opened tabs.

And now it's working on both browsers.

So you were right, it was a sockets issue after all!
